[IV-22-23] Objetivo 2

ignaciotitos opened the pull request:
* [x] ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.
* [x] ¿Los issues representan un problema, y no una tarea?
* [ ] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están asignados a una HU?
* [ ] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit? Los issues sobre los que estoy trabajando, ¿han
      sido asignados al primer milestone?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Son los milestones sobre los que estoy trabajando productos
   mínimamente viables? ¿O tengo que solicitar al product manager que
   cree milestones adicionales o precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

ignaciotitos committed:
Add files via upload

ignaciotitos committed:
Delete AutoIV directory

ignaciotitos committed:
Add files via upload

ignaciotitos commented:
Necesito que asignes los issues al Milestone 1 y este PR al Milestone 0

ignaciotitos commented:
También mira a ver que te parece el código como boceto. Si añadirías algo o no.

RafaelT00 commented:
> Necesito que asignes los issues al Milestone 1 y este PR al Milestone 0

Listo.

> También mira a ver que te parece el código como boceto. Si añadirías algo o no.

Creo que con eso está bien. Al menos para lo que tengo en mente debería bastar.

ignaciotitos commented:
Perfecto. Pues le daré un repaso para marcar todos los puntos y si estás de acuerdo el lunes/martes lo enviamos para que lo corrija JJ.

RafaelT00 commented:
Perfecto.

ignaciotitos commented:
Perdón, necesito que los issues se asignen al milestone 0, perdón por la equivocación.

RafaelT00 commented:
Me parecia raro a mi también, jajaja, bueno ya está listo.

ignaciotitos committed:
Creación del archivo iv.yaml

ignaciotitos commented:
Yo creo que ya estaría. Me ha faltado poner los commits con el formato estándar y de buenas prácticas. Si sabes alguna manera de cambiarlo, lo haré.

RafaelT00 commented:
Creo que puedo solicitar cambios en los commits. Hay dos que simplemente ha sido crear y un directorio y luego borralo no?. Supongo que esos dan igual.

ignaciotitos commented:
Sí. Para añadir la carpeta en los que contiene los ficheros más la creación del fichero iv.yaml que lo cree sin querer para otro pull request.

ignaciotitos committed:
Closes issue #10 porque se crea la clase Empleado

ignaciotitos committed:
Closes issue #12 creando la clase Restricciones

ignaciotitos committed:
Closes issue #11 creando la clase Vacaciones

ignaciotitos commented:
Yo creo que ya estaría para revisión. Tu me dices si ves que hace falta cambiar algo o lo que sea.

ignaciotitos commented:
Cuando lo apruebes mencionamos a JJ.

RafaelT00 reviewed (dismissed):
Me parece correcto.

RafaelT00 commented:
He solicitado cambios sin querer.

ignaciotitos commented:
Perfecto.

danielsp13 commented:
Buenas tardes chicos, he estado haciendo una revisión sobre lo que habéis estado haciendo y tengo que comentaros varios asuntos.

Lo primero de todo, es que ningún *issue* de los planteados está relacionado con alguna de las Historias de Usuario que hay en este proyecto.

En consecuencia de esto, el código que está actualmente implementado no es válido, ya que no se puede verificar si está bien o no.

danielsp13 commented:
Viendo que @RafaelT00 sólo tiene una historia de usuario: #7 . @ignaciotitos , te recomiendo que lo analices y a partir de ahí extraigas lo que se necesita para este objetivo y para satisfacer dicha historia.

Los *issues* deben plantear problemas que se relacionen con dicha HU, más allá de necesito tal atributo o tengo que hacer x clase. La solución que plantees para resolver dicho issue, debe ir en el código. 

Es necesario que sigas este esquema de HU -> Issue -> Solución. Si no, no se te puede dar por bueno este objetivo @ignaciotitos .

danielsp13 commented:
También, se necesita documentar en un fichero el análisis que has realizado, y especifiques por qué has decidido implementar una cosa u otra.

danielsp13 commented:
Finalmente, *commits* como [https://github.com/RafaelT00/AutoIV/pull/13/commits/295218a8bb275099a40c1f8e583c9ab4e9cd40a8](https://github.com/RafaelT00/AutoIV/pull/13/commits/295218a8bb275099a40c1f8e583c9ab4e9cd40a8) no siguen las mejores prácticas. Te recomiendo que también le eches un vistazo a ese asunto.

ignaciotitos commented:
Okey @danielsp13, gracias por tu ayuda. Mañana lo cambiaré. En respecto a lo del commit, sí es verdad que no sigue las mejores prácticas, lo hice sin querer y me di cuenta luego, pero no se como cambiarlo. ¿Tu sabes como?

danielsp13 commented:
> Okey @danielsp13, gracias por tu ayuda. Mañana lo cambiaré. En respecto a lo del commit, sí es verdad que no sigue las mejores prácticas, lo hice sin querer y me di cuenta luego, pero no se como cambiarlo. ¿Tu sabes como?

No te preocupes. A mi también me ha pasado cuando empecé. No es necesario que cambies los commits que tienes ahora mismo. Sólo sigue a partir de ahora con las buenas prácticas y eso será suficiente 😀

danielsp13 commented:
Te lo digo porque manera sí hay... Pero es una mala práctica en general. Así que, no es necesario que lo hagas por ahora siempre y cuando el resto sí que sean correctos.

ignaciotitos committed:
Add new README

Creación de un nuevo README para llevar la documentación de lo que se realiza

ignaciotitos committed:
feat(Empleado.go): Add new lines of code

Añadir antiguedad y vacaciones como variables.

ignaciotitos committed:
Feat(src/README): add documentation

Añadir la documentación sobre objetos valor / entidad y el análisis del dominio.

ignaciotitos commented:
Hola @danielsp13, he cambiado algunas cosas, por lo que si puedes echarle un vistazo cuando puedas, gracias!

danielsp13 commented:
Buenas tardes @ignaciotitos , estoy revisando todo, ahora te comento .

danielsp13 commented on src/README.md:
```diff
@@ -0,0 +1,8 @@
+# Documentación sobre la implementación
+## Análisis del dominio
+Según el [Milestone 0](https://github.com/RafaelT00/AutoIV/milestone/6), para superar este hito es necesario crear una estructura de datos de empleados y otra para los proyectos y las restricciones. Además se necesita una estructura de datos de las vacaciones asignadas. Según la [HU01](https://github.com/RafaelT00/AutoIV/issues/7) se necesita asimilar las peticiones cuando son muchas y asignar va ...
+
+Para esto se han creado issues sobre que estructuras son necesarias, y después de esto, saber que atributos tienen cada clase y además como están conectadas cada una entre las demás.
```
Quizá en vez de hablar directamente sobre el hito, sería mejor que explicaras en la documentación más explícitamente sobre los *issues* y su motivación: por qué se ha planteado el problema o cómo ha surgido, por ejemplo.

ignaciotitos replied:
Vale, perfecto. Gracias por la contestación. A partir de ahora lo haré así.

danielsp13 commented on src/README.md:
```diff
@@ -0,0 +1,8 @@
+
+Para esto se han creado issues sobre que estructuras son necesarias, y después de esto, saber que atributos tienen cada clase y además como están conectadas cada una entre las demás.
+
+## Objeto valor o entidad
+Para cada clase ha sido necesario saber cómo va a funcionar cada clase, para saber si van a ser inmutables o mutables. Para esto ha sido necesario hablar con Rafael y entender cómo se van a asignar las distintas clases. Por ello, las clases Restricciones y Vacaciones, van a ser objetos valor ya que van a ser mutables, mientras que empleado va a ser inmutable, por lo que por eso se va a tener como ...
```
Aquí te has liado un poco. Los objetos valor son aquellos que precisamente, no van a modificar sus datos una vez se han creado.

En cualquier caso, no va mal. Sí que te recomiendo que en esta sección de la documentación, tras aclarar por qué determinada clase es un objeto valor / entidad, especifiques brevemente por qué se soluciona. En los commits se debe contemplar la cadena de cambios hasta alcanzar tu solución propuesta al problema planteado, pero en la documentación también estaría bien indicarlo aunque sea un poco. (para cuando termines de tener todo listo).

ignaciotitos replied:
Ah vale, lo entendí mal. Gracias por la aclaración!

danielsp13 reviewed (commented):
Considero que ahora has hecho un análisis mejor de la situación que se plantea en este problema. Pero es importante que tengas claro la importancia de trabajar con *issues*. 

Los commits que has estado realizando, deberían estar asociados a alguno de los *issues* abiertos para que tanto el autor del repo, como los revisores de este PR, podamos ver mejor para cada uno de ellos cómo los solucionas (el código que propones).

Así que por ejemplo, en commits como [https://github.com/RafaelT00/AutoIV/pull/13/commits/eb6f552e065334a8c9964821a307a944bb2f5fb1](https://github.com/RafaelT00/AutoIV/pull/13/commits/eb6f552e065334a8c9964821a307a944bb2f5fb1), menciona el issue correspondiente. (no obstante, ahora sí siguen mejores prácticas que los anteriores).

ignaciotitos committed:
Update README 

Actualizar el README sobre el análisis del dominio y sobre el objeto valor y la entidad.

ignaciotitos committed:
Update src/README

Actualizo el README con la parte de entidad/objetos valor que faltaba para finalizar la documentación

ignaciotitos commented:
Hola @danielsp13, creo que ya está un poco mejor. Cuando puedas revísalo y con lo que sea me dices. Gracias!

JJ commented:
Por favor, sigue lo comentado por @danielsp13 y lo que hemos hablado en clase.

ignaciotitos committed:
feat: add new entities to iv.yaml

Añadidas las entidades Restricciones y Vacaciones.

ignaciotitos committed:
feat: change the name README to DOC

Cambiar el nombre del archivo de la documentación

ignaciotitos committed:
closes issue #12 invalid 

Issue inválido

ignaciotitos committed:
Remove Restricciones.go because #12 is invalid

ignaciotitos committed:
closes #9 because is invalid

ignaciotitos committed:
closes #10 because is invalid

ignaciotitos committed:
closes #11 because is invalid

ignaciotitos committed:
add: new class Empleado.go

Se añade una nueva clase después de haber eliminado las otras que eran inválidas.

Issue: #17

ignaciotitos committed:
add: new class Vacaciones.go

Se añade la nueva clase Vacaciones después de cerrar la anterior debido a un issue inválido.

Issues: #16 #17
Issue inválido: #11

ignaciotitos committed:
feat: remove entity Restricciones.go

Se borra la entidad Restricciones.go debido a la eliminación de la clase ya que no se va a presentar en el código.

Issue: #12

ignaciotitos committed:
feat: update the class to make it immutable

He corregido la clase para que esta pueda ser un objeto valor y sea inmutable, las condiciones borradas se implementarán dentro de la clase Vacaciones.go

Issue: #15

ignaciotitos committed:
feat: remove the import of Vacaciones.go

Eliminar la importación de la clase Vacaciones.go ya que no se va a necesitar, al menos por ahora, en esta clase.

Issue: #12

ignaciotitos committed:
feat: assignment of an employee on Vacaciones.go

Para poder hacer Empleado un objeto valor he cambiado algunas variables que se pueden establecer en esta clase y las he cambiado para manejarlas aquí.

Issues: #15 #16

ignaciotitos committed:
feat: only one entity

Se han hechos cambios en el código para que sólo Vacaciones sea una entidad y así dejar Empleado como un objeto valor inmutable.

Issue: #16

ignaciotitos commented:
@RafaelT00 aprueba los cambios cuando puedas o sugiere cambios en cualquier cosa.

danielsp13 commented:
Hola @ignaciotitos , por alguna razón que desconozco no he visto nuevos cambios ni menciones hasta ahora. Voy a echarle un vistazo.

danielsp13 commented:
En la documentación `src/DOC.md`:

> Para esto, el primer issue necesario ha sido la https://github.com/RafaelT00/AutoIV/issues/8, en el cual se han propuesto varios tales como Python, Ruby, C++, pero al final el elegido ha resultado ser Go, ya que el compañero quería aprenderlo. Después de asignar el lenguaje me surgió la duda de las https://github.com/RafaelT00/AutoIV/issues/9, que están en el Milestone 0 y se pueden sacar algunas de la HU, pero quería quitarme la duda. Para esto se ha sacado que se necesita de la clase Empleado, Vacaciones y Restricciones.

No me parece un párrafo relevante para lo que se requiere en este aspecto, para eso está el fichero  `iv.yaml`. Salvo que haya un motivo necesariamente importante y que no esté justificado en el correspondiente *issue*, no lo pondría.

****

> Para cada clase ha sido necesario saber cómo va a funcionar cada clase, para saber si van a ser inmutables o mutables. Para esto ha sido necesario hablar con Rafael y entender cómo se van a asignar las distintas clases. Por ello, las clases Restricciones y Vacaciones, van a ser entidades ya que van a ser mutables, mientras que Empleado que iba a ser inmutable de primeras, con los nuevos cambios que se hicieron añadiéndole variables que si van a cambiar durante el ciclo de vida de esta clase cuando se instancie, va a ser otra entidad. En este caso no hay objetos valor, no hay clases inmutables en el tiempo.

> Así con Empleado se soluciona el hecho de crear empleados, ya que es necesario estos para asignar y que les asignen vacaciones. Con la clase Vacaciones se tiene un calendario de los días en los que se puede asignar vacaciones a los empleados, y con la clase Restricciones se soluciona el hecho de que no les puedes asignar vacaciones a los empleados a la vez, tienes que llevar un conteo y poder saber cuando asignar a un empleado o a otro.

Supongo que tendrás que darle una vuelta tras todos estos cambios que has ido realizando.

danielsp13 reviewed (commented):
Los commits son bastante mejores que los que habías hecho al principio, aunque si que creo que hay determinadas descripciones que son innecesarias: [https://github.com/RafaelT00/AutoIV/pull/13/commits/cea9a0adc6d115aeffa16d1001ea8de95b683fd2](https://github.com/RafaelT00/AutoIV/pull/13/commits/cea9a0adc6d115aeffa16d1001ea8de95b683fd2), ya se ve claramente lo que es con el cuerpo, y con los cambios realizados.

Con respecto a todo lo demás, creo que ahora sí que se ha seguido la metodología y la solución que planteas considero que es la que se requiere según los issues que planteaste (por lo menos, se acerca).

Asegúrate de realizar toques finales especialmente en la documentación, salvo que @RafaelT00 te comente algo diferente por supuesto.

ignaciotitos committed:
feat: remove some innecesary stuff about the doc

ignaciotitos committed:
feat: add new explanation about Empleado

ignaciotitos commented:
Vale. Cambiado. Gracias por la ayuda @danielsp13.

ignaciotitos commented:
@RafaelT00 cuando puedas aprueba o sugiere cambios.

ignaciotitos committed:
add: new class to carry restrictions 

Clase para llevar las restricciones con los empleados y poder así asignar las vacaciones.

Issue: #16

ignaciotitos committed:
feat: new entity in modelling

ignaciotitos committed:
feat: correct the syntaxis for YAML

ignaciotitos committed:
feat: correct bad syntax for YAML

RafaelT00 reviewed (commented):
Me parece correcto.

ignaciotitos commented:
Tienes que aprobar los cambios @RafaelT00

RafaelT00 reviewed (approved):
Todo bien.

ignaciotitos commented:
Listo para revisión @JJ

JJ commented:
Ni un solo commit tiene el issue al que se refiere en primera línea. Eso aumenta el trabajo que tengo que hacer, pulsando en cada uno para ver si está en el resto del cuerpo. Os agradecería que siguiérais las buenas prácticas repetidas en clase y en los errores frecuentes, porque todas tienen su razón de ser.

JJ commented:
> > Necesito que asignes los issues al Milestone 1 y este PR al Milestone 0
> 
> Listo.
> 
> > También mira a ver que te parece el código como boceto. Si añadirías algo o no.
> 
> Creo que con eso está bien. Al menos para lo que tengo en mente debería bastar.

Es casi imposible que esté bien; siempre habrá alguna cosa que no corresponda a las historias de usuario o que no encaje del todo. Te recuerdo que no se trata de que @ignaciotitos averigüe lo que tienes en mente, sino que los dos tratéis de modelizar correctamente, siguiendo en lo posible *domain driven design* el problema. Si no, es dejarme la labor de evaluar a mi (lo tengo que hacer de todas formas y lo hago con gusto, pero en la revisión y corrección *los dos* aprenderéis algo)

JJ commented on src/DOC.md:
```diff
@@ -0,0 +1,14 @@
+# Documentación sobre la implementación
+## Análisis del dominio
+Me surgió la necesidad de preguntar los atributos que iban a tener cada clase, Empleado, Vacaciones y Restricciones. 
```
Ya partes de un error. No se pueden *preguntar* los atributos. Tienes que plantear el problema, como se supone que has hecho porque has marcado la casilla correspondiente en el cuerpo.

JJ commented on src/DOC.md:
```diff
@@ -0,0 +1,14 @@
+# Documentación sobre la implementación
```
Un directorio llamado `src` *nunca* es el mejor sitio para la documentación. Debes seguir las buenas prácticas aquí también.

JJ reviewed (changes requested):
Partiendo de esto:
![Captura de pantalla de 2022-12-13 14-21-35](https://user-images.githubusercontent.com/500/207329597-137916f7-dc0a-459e-bd2a-e88b2cafcdee.png)
el resto no puede estar bien. A estas alturas de la asignatura *no puedes usar la web* para desarrollar, sino la línea de órdenes o el IDE. Debes aprender a usar las herramientas más adecuadas para el desarrollo. Por no mencionar que los mensajes de commit deben seguir las buenas prácticas.
Por favor, revisa los objetivos de aprendizaje de las últimas semanas, repasa si de verdad estás haciendo lo que dice en la lista de comprobación, y los errores frecuentes de este objetivo.

ignaciotitos commented:
> Ni un solo commit tiene el issue al que se refiere en primera línea. Eso aumenta el trabajo que tengo que hacer, pulsando en cada uno para ver si está en el resto del cuerpo. Os agradecería que siguiérais las buenas prácticas repetidas en clase y en los errores frecuentes, porque todas tienen su razón de ser.

Busqué por internet en varios sitios y ponían el issue al final del todo.

ignaciotitos committed:
feat: remove innecesary and bad stuff #9

JJ commented:
> > Ni un solo commit tiene el issue al que se refiere en primera línea. Eso aumenta el trabajo que tengo que hacer, pulsando en cada uno para ver si está en el resto del cuerpo. Os agradecería que siguiérais las buenas prácticas repetidas en clase y en los errores frecuentes, porque todas tienen su razón de ser.
> 
> Busqué por internet en varios sitios y ponían el issue al final del todo.

¿Entiendes por qué hace falta que el issue esté en la primera línea, y por qué lo he dicho así repetidamente en clase, y repetido también en la sección de "errores frecuentes" de alguna semana, y en las presentaciones que he hecho? Se trata de que entendáis por qué hacéis las cosas. "Buscar por internet" no es justificación suficiente (ninguna justificación, de hecho) para no seguir las mejores prácticas, cuando además se te ha dado la razón por la que se ha pide que se haga así.

ignaciotitos committed:
docs: change directory and name

ignaciotitos commented:
Ya cambié el path del documento DOC.md desde Git Bash y le cambié el nombre al fichero.

ignaciotitos commented:
> > > Ni un solo commit tiene el issue al que se refiere en primera línea. Eso aumenta el trabajo que tengo que hacer, pulsando en cada uno para ver si está en el resto del cuerpo. Os agradecería que siguiérais las buenas prácticas repetidas en clase y en los errores frecuentes, porque todas tienen su razón de ser.
> > 
> > 
> > Busqué por internet en varios sitios y ponían el issue al final del todo.
> 
> ¿Entiendes por qué hace falta que el issue esté en la primera línea, y por qué lo he dicho así repetidamente en clase, y repetido también en la sección de "errores frecuentes" de alguna semana, y en las presentaciones que he hecho? Se trata de que entendáis por qué hacéis las cosas. "Buscar por internet" no es justificación suficiente (ninguna justificación, de hecho) para no seguir las mejores prácticas, cuando además se te ha dado la razón por la que se ha pide que se haga así.

A partir de ahora lo haré así.

ignaciotitos commented:
> Partiendo de esto: ![Captura de pantalla de 2022-12-13 14-21-35](https://user-images.githubusercontent.com/500/207329597-137916f7-dc0a-459e-bd2a-e88b2cafcdee.png) el resto no puede estar bien. A estas alturas de la asignatura _no puedes usar la web_ para desarrollar, sino la línea de órdenes o el IDE. Debes aprender a usar las herramientas más adecuadas para el desarrollo. Por no mencionar que los mensajes de commit deben seguir las buenas prácticas. Por favor, revisa los objetivos de aprendizaje de las últimas semanas, repasa si de verdad estás haciendo lo que dice en la lista de comprobación, y los errores frecuentes de este objetivo.

Entiendo. Me pareció en el momento la forma más fácil y rápida de hacerlo. A partir de ahora lo haré con Git Bash que ya he aprendido y me puedo defender un poco.

ignaciotitos commented:
@JJ para revisar otra vez.

JJ commented on docs/documentacion_obj2.md:
```diff
@@ -0,0 +1,10 @@
+Después tuve que ver si había alguna [diferencia entre Empleado y el encargado de asignar vacaciones](https://github.com/RafaelT00/AutoIV/issues/17), en el cuál quedo claro que no había falta diferenciar entre estos dos.
+
+## Objeto valor o entidad
+Para cada clase ha sido necesario saber cómo va a funcionar cada clase, para saber si van a ser inmutables o mutables. Para esto ha sido necesario hablar con Rafael y entender cómo se van a asignar las distintas clases. Por ello, la clase Vacaciones, va a ser una entidad ya que van a ser mutables. Empleado en este caso se han hecho unas modificaciones para que las partes mutables de Empleado se t ...
```
Como pone en los problemas frecuentes expresados en diferentes sesiones, esto no es una sesión de preguntas y respuestas. La persona que se encargue de programar tiene que plantear los problemas y proponer soluciones, y el único diálogo es a través del código en este PR.

JJ replied:
... y si se ha hecho así, es posible que no se haya seguido la metodología correcta y por tanto el resultado sea incorrecto.
También: si hacemos un comentario es para que reviséis la línea en la que se hace.

JJ commented on src/Empleado.go:
```diff
@@ -0,0 +1,10 @@
+
+type Empleado struct{
+	Nombre string
+	Ficha string
```
¿Qué tiene que ver la ficha con las vacaciones?

ignaciotitos replied:
Es la ficha del empleado, para identificarlo

JJ commented on src/Restricciones.go:
```diff
@@ -0,0 +1,13 @@
+import (
+	fmt
+	time
+  Empleado
```
Por favor, formatea correctamente el código.

ignaciotitos replied:
Hecho

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+lenguaje: Go
+entidad: 
+  - src/Vacaciones.go
```
Por favor, infórmate de las mejores prácticas para los directorios donde tienen que estar los fuentes en Go. No parece que `src` sea el sitio habitual.

JJ commented on src/Vacaciones.go:
```diff
@@ -0,0 +1,18 @@
+type Vacaciones struct{
+	Dias uint
+	Periodo []time.Time
+	type Worker struct{
```
¿Hay un `Empleado` y también un `Worker`?

JJ commented on src/Restricciones.go:
```diff
@@ -0,0 +1,13 @@
+)
+
+type Restricciones struct{
+	NombreProyecto string
```
En #7 no aparece por ningún lado la palabra "proyecto". Si no aparece, es que no forma parte del milestone.

ignaciotitos replied:
A mi entender las HUs son pequeñas tareas para un milestone en concreto, si entiendo correctamente. Por lo que si pertenece al milestone pero no a esa HU, ¿no?

JJ replied:
No, no lo son. Repasa lo impartido sobre el tema, presentaciones, objetivos de aprendizaje.

JJ reviewed (changes requested):
No hay ningún issue que plantee los problemas que hay en esta aplicación, o que se planteen a partir de #7. Por otro lado, la misma #7 debería aclararse un poco más, porque las historias de usuario son el lugar donde se tiene que plantear claramente la necesidad del usuario, no contestaciones a issues (de las que no debe haber ninguna)

ignaciotitos committed:
change: rehacer Vacaciones según #7

ignaciotitos committed:
change: rehacer Vacaciones según #7

ignaciotitos committed:
change: cambiar directorio src por internal

ignaciotitos committed:
change: eliminar empleados por innecesario #7

ignaciotitos commented:
He hecho algunos cambios en los que creo que ha podido quedar un poco más claro. @RafaelT00 @JJ

ignaciotitos committed:
docs: cambiar algunas líneas de la documentación

ignaciotitos commented:
Listo para revisión. @JJ

JJ commented on internal/Restricciones.go:
```diff
@@ -0,0 +1,10 @@
+
+import "time"
+
+type Restricciones struct {
```
No se puede desarrollar código directamente de una historia de usuario. Hay que crear una serie de issues para poder ir dividiendo el problema en diferentes partes, ver la relación entre las mismas y ver qué se va a usar en la lógica de negocio o no.
Esto está desarrollado en un commit https://github.com/RafaelT00/AutoIV/commit/5d82eb336ae4bf2b8760298bc380af5c8aecc6b5 que apunta directamente a #7 (que no tiene título, por cierto), sin crear ningún issue intermedio. Es imposible de evaluar, porque no se puede saber si se han resuelto los diferentes problema que existen, simplemente porque ni siquiera se han planteado los problemas en su issue correspondiente.

JJ reviewed (changes requested):
Es imposible llegar a una solución correcta si no se sigue la metodología centrada en el cliente, con los pasos intermedios adecuados. Este milestone no es "programar cosas que tengan nombres mencionados en una historia usuario", es plantear una serie de problemas y crear, a partir de ellos, un modelo del problema. Por favor, repasad el material correspondiente enlazado desde las diferentes sesiones y los errores frecuentes.

ignaciotitos committed:
closes #12 por no suponer nada en #7

ignaciotitos committed:
closes #10 por mala implementación del problema

ignaciotitos committed:
closes #11 por mala implementación del problema

ignaciotitos committed:
add: resuelve problema relacionado con #19

ignaciotitos committed:
add: solucion propuesta al issue #20 y #21

ignaciotitos committed:
add: solución propuesta al issue #21

ignaciotitos commented:
@RafaelT00 solicito revisión.

ignaciotitos committed:
actualizar iv.yaml después de los cambios

ignaciotitos committed:
change: #21 close #17 #16 #15 #18

ignaciotitos committed:
Merge branch 'objetivo-2' of https://github.com/ignaciotitos/AutoIV into objetivo-2

RafaelT00 reviewed (commented):
Encargado creo que debería ser simplemente un empleado. El resto lo veo bien.

ignaciotitos commented:
Pero entonces como va a asignar las vacaciones? Ya que habría que diferenciar entre un empleado y el encargado que asigna las vacaciones.

RafaelT00 commented:
Es el software el que asigna las vacaciones a los empleados. Los encargados son los que lo usarían, quiero decir, no se si tiene sentido representar a quien usa el software.

ignaciotitos commented:
Vale, entiendo. Cambio de Encargado a Empleado

ignaciotitos committed:
close #19 y se actualiza el nombre #22

ignaciotitos committed:
feat: mala sintaxis del código

ignaciotitos committed:
change: mala sintaxis al declarar la clase

ignaciotitos committed:
change: mala sintaxis y mala práctica en go

ignaciotitos commented:
Ya estaría. Si te parece bien aprueba los cambios y menciono al profesor.

RafaelT00 reviewed (approved):
Me parece bien.

ignaciotitos committed:
docs: añadir nuevos cambios en la documentación

ignaciotitos committed:
docs: finalizar la documentación

ignaciotitos commented:
Listo para revisión @JJ

JJ commented on internal/Empleado.go:
```diff
@@ -0,0 +1,8 @@
+
+type Empleado struct {
+	Ficha  string
+	Asigna []time.Time
```
` Asigna` es un verbo, no un sustantivo.

ignaciotitos replied:
Fallo mío.

JJ commented on docs/documentacion_obj2.md:
```diff
@@ -0,0 +1,12 @@
+
+Después el siguiente problema que me encontré son las tareas, en las que tengo que comprender como cuantificar las tareas importantes, cual es su dificultad, cuantas tareas tiene un empleado. Planteado en RafaelT00#20. En este problema planteo una solución que la propongo con este commit [17f2199](https://github.com/ignaciotitos/AutoIV/commit/17f2199aa10ca01182bf7c21eb058762507fe183)
+
+Más tarde surgió el problema de como se iban a gestionar las peticiones por parte de los encargados, cuando son las épocas de mayores peticiones y como afectan a las asignaciones. Este problema está referenciado en este issue RafaelT00#21, en el que añadí una solución en los commmits 
```
Creo que no lo habéis entendido, una vez más. No pueden "surgir" "más tarde". Esto es el enemigo de la eficiencia y de la coherencia en el código. Las historias de usuario deben ser todo lo que hace falta para programar, junto con los milestones, que dicen hasta qué punto hay que hacerlo.

ignaciotitos replied:
Esto lo digo de como he ido desentrañando la historia de usuario frase a frase para encontrarme los diversos problemas que se proponen en esta para luego proponer una solución.

JJ reviewed (changes requested):
Para empezar, los issues tienen que estar asignados a un milestone. Hay que centrarse, también, en un milestone determinado, y hacer un producto mínimamente viable. Esto debe ser un modelo coherente de al menos parte del dominio del problema, partiendo de las historias de usuario.
También conviene seguir buenas prácticas mínimas sobre nombres de variables.

ignaciotitos committed:
change: cambiar verbo por sustantivo

ignaciotitos commented:
> Para empezar, los issues tienen que estar asignados a un milestone. Hay que centrarse, también, en un milestone determinado, y hacer un producto mínimamente viable. Esto debe ser un modelo coherente de al menos parte del dominio del problema, partiendo de las historias de usuario. También conviene seguir buenas prácticas mínimas sobre nombres de variables.

Fallo mío, no me acordé se decírselo al compañero. Ya está resuelto.

ignaciotitos committed:
change: seguir buenas prácticas de código

ignaciotitos committed:
change: seguir buenas prácticas de código en go

ignaciotitos committed:
change: seguir buenas prácticas en Peticiones

ignaciotitos committed:
docs: cambio la documentación

ignaciotitos commented:
Listo para revisión @JJ

JJ reviewed (changes requested):
Se ve que me dejé esto. En todo caso, seguimos en lo mismo: hay que hacer un modelo del problema.

JJ commented on docs/documentacion_obj2.md:
```diff
@@ -0,0 +1,12 @@
+# Documentación sobre la implementación
+## Análisis del dominio
+Analizamos la HU RafaelT00#7. Es necesario entender como se van a relacionar las distintas partes ahí escritas. Primero para dividir la historia de usuario en trozos más pequeños para entenderlos y luego relacionarlos, obtengo los encargados de asignar las tareas, en los que me surge un problema de como contextualizarlos RafaelT00#19, en el cual creo una solución que se puede adaptar [ecbb97e](ht ...
```
Pon un enlace. En estos documentos no se enlaza automáticamente como en los issues.

ignaciotitos replied:
Ahí está en enlace.

JJ replied:
Sigue sin estar el enlace al issue.

JJ commented on docs/documentacion_obj2.md:
```diff
@@ -0,0 +1,12 @@
+## Análisis del dominio
+Analizamos la HU RafaelT00#7. Es necesario entender como se van a relacionar las distintas partes ahí escritas. Primero para dividir la historia de usuario en trozos más pequeños para entenderlos y luego relacionarlos, obtengo los encargados de asignar las tareas, en los que me surge un problema de como contextualizarlos RafaelT00#19, en el cual creo una solución que se puede adaptar [ecbb97e](ht ...
+
+El siguiente problema en RafaelT00#7 que me encontré son las tareas, en las que tengo que comprender como cuantificar las tareas importantes, cual es su dificultad, cuantas tareas tiene un empleado. Planteado en RafaelT00#20. En este problema planteo una solución que la propongo con este commit [17f2199](https://github.com/ignaciotitos/AutoIV/commit/17f2199aa10ca01182bf7c21eb058762507fe183)
```
¿por qué es importante esto para las vacaciones? Los problemas tienes que plantearlos en un issue independiente. En este caso, además, decidir qué relevancia tiene o no.

JJ replied:
:arrow_up:

JJ commented on docs/documentacion_obj2.md:
```diff
@@ -0,0 +1,12 @@
+
+El siguiente problema en RafaelT00#7 que me encontré son las tareas, en las que tengo que comprender como cuantificar las tareas importantes, cual es su dificultad, cuantas tareas tiene un empleado. Planteado en RafaelT00#20. En este problema planteo una solución que la propongo con este commit [17f2199](https://github.com/ignaciotitos/AutoIV/commit/17f2199aa10ca01182bf7c21eb058762507fe183)
+
+El problema siguiente en RafaelT00#7 fue de como se iban a gestionar las peticiones por parte de los encargados, cuando son las épocas de mayores peticiones y como afectan a las asignaciones. Este problema está referenciado en este issue RafaelT00#21, en el que añadí una solución en los commmits 
```
Insisto, ¿qué tiene que ver esto con las vacaciones?

ignaciotitos replied:
Entiendo que según las peticiones que le lleguen dependerán los días que les asignen, si las peticiones depende de la franja del año y cuántos días ha pedido de vacaciones.

JJ replied:
1. Sigues sin enlazar correctamente los issues.
2. Sigue sin tener relevancia el "encargado" y lo que este haga.

JJ commented on internal/Empleado.go:
```diff
@@ -0,0 +1,8 @@
+
+type Empleado struct {
+	ficha  	             string
+	asignacionVacaciones []time.Time
```
¿De qué forma esto sería una característica de un empleado y no un resultado de una función?
En general, usar un campo de un struct o un atributo para cada palabra que aparezca en los HUs no es una buea idea. Lo es plantear los diferentes problemas que partan de ahí, y resolverlos.

JJ reviewed (changes requested):
No he visto un cambio sustancial, pero lo principal es que no ha habido un cambio en la metodología necesaria para superar este objetivo.

ignaciotitos committed:
change: cambio de nombre del fichero

ignaciotitos committed:
close #23 y cambio de nombre y rehacer la clase

ignaciotitos committed:
close #22 y quitar parte del código inservible

ignaciotitos committed:
change: se cambia el nombre del archivo

ignaciotitos committed:
close #24 al proponer una solución

ignaciotitos committed:
close #25 al proponer una solución para proyectos

ignaciotitos commented:
Listo para revisión @JJ

ignaciotitos committed:
docs: agregar los nuevos problemas encontrados

JJ commented on internal/Peticiones.go:
```diff
@@ -0,0 +1,32 @@
+package internal
+
+type Season uint8
```
¿Una estación es un entero de 8 bits? ¿No será una estación?

JJ commented:
Por favor, seguid las mejores prácticas en mensajes de commit
1. https://github.com/RafaelT00/AutoIV/commit/7a24cff122951ebfdc01d8bd84dbcda3713af3af no dice por qué propone una solución ni qué solucion propone. https://github.com/RafaelT00/AutoIV/pull/13/commits/87b2b13db1db49609e085fe3556374a2cf9dc3ab igual. No es correcto no seguir buenas prácticas, es incluso peor hacerlo varias veces.
2. 1cb5dcfa99708c447a56e00d0defed22454859e1 borra un archivo y no dice por qué lo hace; además, dice que lo "cambia de nombre". Cambiar de nombre se hace con `git mv`. Y siempre se indica por qué, y qué problema o issue resuelve.

JJ commented on internal/Peticiones.go:
```diff
@@ -0,0 +1,32 @@
+}
+
+// s.String() devuelve el dia de la semana en string en vez de en numero
+func (s Semana) String() string {
```
De veras que llegado a este punto estoy totalmente perdido. Ni entiendo qué relevancia pueden tener los nombres de los días de la semana, ni por qué usáis una función y no una simple estructura de datos, ni nada.

JJ commented on internal/Proyecto.go:
```diff
@@ -0,0 +1,16 @@
+type Tarea uint8
+
+const (
+	Facil   Tarea = 0
```
Esto continúa aquí, a pesar de que hablamos ayer de que no había nada en la historia de usuario que apoyara su existencia.

ignaciotitos replied:
En la HU pone que los encargados tienen tareas más importantes que hacer, lo que los empleados también tienen que hacer tareas relacionadas con el proyecto. Con esto se ve si una tarea es importante o no, para poder asignar más o menos días de vacaciones.

JJ replied:
No, no se ve ni eso es relevante a la lógica de negocio. ¿De qué forma influye que el encargado tenga una tarea de dificultad "muy difícil" para que yo pida 3 días de vacaciones en abril? Os rogaría que para optimizar el tiempo de todo el mundo implicado, tratárais de entender un poco mejor tanto el problema que tratáis como las objeciones que hacemos a vuestro trabajo relacionado con el mismo.

ignaciotitos replied:
En el momento que lo pensé tiene bastante sentido ya que se le asigna un número, en el que va a tener una serie de tareas que van a poder multiplicar de manera mayor o menor el hecho de que las tareas sean más o menos complicadas ya que eso requerirá más o menos tiempo.

JJ commented on internal/Proyecto.go:
```diff
@@ -0,0 +1,16 @@
+type Proyecto struct {
+	tareas           []Tarea
+	empleadosEnTarea []Empleado
+	durabilidad      uint8
```
¿Qué es la durabilidad? ¿Por qué es importante en el contexto de la lógica de negocio?

ignaciotitos replied:
Es la duración del proyecto. Lo cambio ahora para que quede más claro.

JJ replied:
¿De qué forma la duración de un proyecto influye en las vacaciones que pida alguien? Os agradecería que planteárais los problemas correctamente, para poder llegar a soluciones válidas.

ignaciotitos replied:
Para saber si queda poco tiempo de proyecto y le quedan por hacer muchas tareas pues no se le pueden asignar vacaciones por ejemplo.

JJ replied:
¿Ese problema lo has planteado en qué issue?

JJ reviewed (changes requested):
Como os he comentado repetidamente, el proceso es importante, y este pide que empecéis por las historias de usuario y vayáis entendiendo qué es necesario para construir la lógica de negocio que proporcione una solución a ese usuario, a base de dividir el problema y hallar soluciones parciales. Como os he comentado en varios issues, no hay ninguna evidencia de que se haya hecho así, ni de que se sigan buenas prácticas en los mensajes de commit, que evidencian que no se entiende la función (fundamental) de los mismos ni la que cumplen los issues en relación a la metodología de resolución de problemas que es inherente a la ingeniería informática.
Sería conveniente que repasárais el material sobre desarrollo ágil, el significado de un issue y de las historias de usuario, y las buenas prácticas en creación y manejo de las mismas.

ignaciotitos commented:
> Por favor, seguid las mejores prácticas en mensajes de commit
> 
> 1. [7a24cff](https://github.com/RafaelT00/AutoIV/commit/7a24cff122951ebfdc01d8bd84dbcda3713af3af) no dice por qué propone una solución ni qué solucion propone. [87b2b13](https://github.com/RafaelT00/AutoIV/commit/87b2b13db1db49609e085fe3556374a2cf9dc3ab) igual. No es correcto no seguir buenas prácticas, es incluso peor hacerlo varias veces.
> 2. [1cb5dcf](https://github.com/RafaelT00/AutoIV/commit/1cb5dcfa99708c447a56e00d0defed22454859e1) borra un archivo y no dice por qué lo hace; además, dice que lo "cambia de nombre". Cambiar de nombre se hace con `git mv`. Y siempre se indica por qué, y qué problema o issue resuelve.

Fallo mío. A partir de ahora lo haré así.

ignaciotitos committed:
change: reformular código para ser más entendible

ignaciotitos committed:
change: cambio nombre variable #25

ignaciotitos committed:
docs: actualizar la documentación

ignaciotitos commented:
Listo para revisión @JJ

JJ commented:
> Listo para revisión @JJ

De la misma forma, si no habéis entendido algunos de los comentarios y por tanto no habéis actuado al respecto, os rogaría que esperárais a mi contestación a vuestras objeciones para volver a enviarlo. De esa forma podremos todos usar mejor nuestro tiempo.

ignaciotitos commented:
> > Listo para revisión @JJ
> 
> De la misma forma, si no habéis entendido algunos de los comentarios y por tanto no habéis actuado al respecto, os rogaría que esperárais a mi contestación a vuestras objeciones para volver a enviarlo. De esa forma podremos todos usar mejor nuestro tiempo.

Entiendo los comentarios, lo que no entiendo es porque tantos inconvenientes. He puesto los problemas que me han ido surgiendo entendiendo lo que se quiere hacer, la HU y el milestone 0. Lo mismo es por cómo se resuelven los problemas en lo que nunca estaremos de acuerdo ya que no puedo mirar ni pensar como usted. Ya no se como atajar este objetivo la verdad.

JJ commented:
> > > Listo para revisión @JJ
> > 
> > 
> > De la misma forma, si no habéis entendido algunos de los comentarios y por tanto no habéis actuado al respecto, os rogaría que esperárais a mi contestación a vuestras objeciones para volver a enviarlo. De esa forma podremos todos usar mejor nuestro tiempo.
> 
> Entiendo los comentarios, lo que no entiendo es porque tantos inconvenientes. He puesto los problemas que me han ido surgiendo entendiendo lo que se quiere hacer, la HU y el milestone 0. Lo mismo es por cómo se resuelven los problemas en lo que nunca estaremos de acuerdo ya que no puedo mirar ni pensar como usted. Ya no se como atajar este objetivo la verdad.

Pues conviene que lo entiendas. Te sugiero que sigas el material impartido sobre domain driven design, que comprendas la importancia de las historias de usuario y de plantear los issues como problemas y repases los errores frecuentes en el objetivo 2 expresados a lo largo de las semanas.
